client/src/glossary/GlossarySidebar/GlossarySidebarController.tsx

```diff
@@ -0,0 +1,170 @@
+  glossary_item_key: string;
+  focus_item_key: string;
+  return_focus_target: HTMLElement | undefined;
+  search_phrase: string;
+  keydown_close: boolean;
+}
+
+const GlossarySidebarController = withRouter(
+  class GlossarySidebarController extends React.Component<
+    RouteComponentProps,
+    GlossarySidebarControllerState
+  > {
+    menu_ref = React.createRef<HTMLDivElement>();
+
+    constructor(props: RouteComponentProps) {
+      super(props);
+      this.state = {
+        is_open: false,
+        glossary_item_key: "",
+        focus_item_key: "",
+        return_focus_target: undefined,
+        search_phrase: "",
+        keydown_close: false,
+      };
+    }
+    itemClick = (e: HTMLElement) => {
+      const target = e;
+      const glossary_item_key = target.dataset.ibttGlossaryKey;
+
+      if (glossary_item_key) {
+        this.setGlossaryItem(glossary_item_key);
+      }
+
+      this.setState({
+        return_focus_target: e,
+      });
+
+      this.toggleGlossary(true);
+    };
+
+    handleWindowClick = (e: Event) => {
+      const target = (e.target as HTMLElement).closest(
+        "[data-toggle=glossary_sidebar]"
+      );
+      if (target) {
+        this.itemClick(target as HTMLElement);
+      }
+    };
+
+    handleWindowKeyDown = (e: KeyboardEvent) => {
+      if (e.key === "Enter") {
+        this.handleWindowClick(e);
+      }
+    };
+
+    componentDidMount() {
+      window.addEventListener("click", this.handleWindowClick, {
+        capture: true,
+      });
+      window.addEventListener("keydown", this.handleWindowKeyDown, {
+        capture: true,
+      });
+    }
+
+    componentWillUnmount() {
+      window.removeEventListener("click", this.handleWindowClick, {
+        capture: true,
+      });
+      window.removeEventListener("keydown", this.handleWindowKeyDown, {
+        capture: true,
+      });
+    }
+
+    toggleGlossary(value: boolean) {
+      this.setState({
+        is_open: value,
+        keydown_close: false,
+      });
+    }
+
+    keydownClose() {
+      this.setState({
+        is_open: false,
+        keydown_close: true,
+      });
+    }
+
+    setGlossaryItem(key: string) {
+      this.setState({
+        glossary_item_key: key,
+        focus_item_key: key ? key : this.state.focus_item_key,
+      });
+    }
+
+    setQuery(query: string) {
+      this.setState({
+        search_phrase: query,
+      });
+    }
+
+    buttonClick() {
+      this.toggleGlossary(true);
+      this.setState({
+        return_focus_target: undefined,
+        keydown_close: true,
+      });
+    }
+
+    render() {
+      const currentPage = this.props.location.pathname;
+
+      if (routes_without_glossary.includes(currentPage)) return null;
+
+      return (
+        <>
+          <GlossarySidebar
+            glossary_item_key={this.state.glossary_item_key}
+            focus_item_key={this.state.focus_item_key}
+            set_glossary_item={(key) => this.setGlossaryItem(key)}
```

Comment:
The old advice (still applies in classes, but function components with hooks just give up on it far as I know) is to avoid call backs and event handlers that are declared inside a component's `render`, so that React can tell when those props actually are changing. Inlined arrow functions are re-declared on each render! The pattern here should just be `set_glossary_item={this.setGlossaryItem}` etc.

Reply:
This occurs across other files in this PR, so (as always) do your own pass after addressing the specific things I've pointed out in this review round :+1:

Reply:
Can discuss this in our dev meeting, I have a few questions, thanks!